packages/blade/src/components/Toast/ToastContainer.web.tsx

```diff
@@ -0,0 +1,271 @@
+  const { toasts, handlers } = useToaster(toastOptions);
+  const { theme } = useTheme();
+  const [frontToastHeight, setFrontToastHeight] = React.useState(0);
+  const [hasManuallyExpanded, setHasManuallyExpanded] = React.useState(false);
+  const isMobile = useIsMobile();
+  const MIN_TOASTS = isMobile ? 1 : 3;
+  const CONTAINER_OFFSET = isMobile ? 16 : 24;
+
+  const infoToasts = React.useMemo(() => toasts.filter((toast) => !isPromotionalToast(toast)), [
+    toasts,
+  ]);
+  const promoToasts = React.useMemo(() => toasts.filter((toast) => isPromotionalToast(toast)), [
+    toasts,
+  ]);
+
+  // always keep promo toasts at the bottom of the stack
+  const recomputedToasts = React.useMemo(() => [...infoToasts, ...promoToasts], [
+    infoToasts,
+    promoToasts,
+  ]);
+
+  const hasPromoToast = promoToasts.length > 0 && promoToasts[0]?.visible;
+  const promoToastHeight = promoToasts[0]?.height ?? 0;
+  const isExpanded = hasManuallyExpanded || recomputedToasts.length <= MIN_TOASTS;
+
+  React.useEffect(() => {
+    if (hasManuallyExpanded) {
+      handlers.startPause();
+    } else {
+      handlers.endPause();
+    }
+  }, [handlers, hasManuallyExpanded]);
+
+  React.useLayoutEffect(() => {
+    // find the first toast which is visible
+    setFrontToastHeight(infoToasts.find((t, index) => t.visible && index === 0)?.height ?? 0);
+  }, [infoToasts]);
+
+  // calculate total height of all toasts
+  const totalHeight = React.useMemo(() => {
+    return (
+      recomputedToasts
+        // only consider visible recomputedToasts
+        .filter((toast) => toast.visible)
+        .reduce((prevHeight, toast) => prevHeight + (toast.height ?? 0), 0) +
+      recomputedToasts.length * CONTAINER_OFFSET
+    );
+  }, [recomputedToasts, CONTAINER_OFFSET]);
+
+  const calculateYPosition = React.useCallback(
+    ({ toast, reverseOrder = false, index, defaultPosition }: CalculateYPositionProps) => {
+      const relevantToasts = infoToasts.filter(
+        (t) => (t.position ?? defaultPosition) === (toast.position ?? defaultPosition) && t.height,
+      );
+      const toastIndex = relevantToasts.findIndex((t) => t.id === toast.id);
+      // number of toasts before this toast
+      const toastsBefore = relevantToasts.filter((toast, i) => i < toastIndex && toast.visible)
+        .length;
+
+      let scale = index < MAX_TOASTS ? 1 : Math.max(0.7, 2 - (toastsBefore * SCALE_FACTOR + 1));
+      // y position of toast,
+      let offset = relevantToasts
+        .filter((toast) => toast.visible)
+        .slice(...(reverseOrder ? [toastsBefore + 1] : [0, toastsBefore]))
+        .reduce((acc, toast) => {
+          if (isExpanded) {
+            return acc + (toast.height ?? 0) + GUTTER;
+          }
+          return acc + PEEK_GUTTER;
+        }, 0);
+
+      // lift all info toasts up if there is a promo toast
+      if (hasPromoToast) {
+        offset += GUTTER + promoToastHeight;
+      }
+      // promo toasts should always be on bottom
```

Comment:
We need to better document all of this.

1. Why are we doing this?
2. What this is doing?

Reply:
Refactored the code a bit, added some comments and a link to video explanting the stacking code: 

https://www.loom.com/share/522d9a445e2f41e1886cce4decb9ab9d?sid=4287acf6-8d44-431b-93e1-c1a0d40a0aba